File: src/info/meodinger/LabelPlusFX/Component/CColorPicker.java
package info.meodinger.LabelPlusFX.Component;

import info.meodinger.LabelPlusFX.Util.CColor;
import info.meodinger.LabelPlusFX.Util.CString;

import javafx.beans.property.StringProperty;
import javafx.event.ActionEvent;
import javafx.geometry.Pos;
import javafx.scene.Node;
import javafx.scene.control.*;
import javafx.scene.control.TextField;
import javafx.scene.control.skin.ColorPickerSkin;
import javafx.scene.input.KeyCode;
import javafx.scene.layout.Region;
import javafx.scene.layout.VBox;
import javafx.scene.paint.Color;

import java.util.List;
import java.util.stream.Collectors;


/**
 * @author Meodinger
 * Date: 2021/5/28
 * Location: info.meodinger.LabelPlusFX.Component
 */
public class CColorPicker extends ColorPicker {

    private final TextField colorHexField;

    public CColorPicker() {
        super();
        this.colorHexField = new TextField();
        init();
    }

    public CColorPicker(Color color) {
        this();
        setValue(color);
    }

    private void init() {

        colorHexField.setTextFormatter(new TextFormatter<String>(change -> {
            change.setText(change.getText().toUpperCase());
            if (change.isAdded()) {
                StringBuilder builder = new StringBuilder();
                for (char c : change.getText().toCharArray()) {
                    if ((c >= '0' && c <= '9') || (c >= 'A' && c <= 'F')) {
                        builder.append(c);
                    }
                }
                change.setText(builder.toString());
            }
            return change;
        }));
        colorHexField.setOnKeyReleased(event -> {
            if (event.getCode() == KeyCode.ENTER) {
                this.fireEvent(new ActionEvent());
                this.hide();
                // Though enter pressed and picker hide, but if not invoke hide(), picker will not show again
            }
        });

        setOnShown(event -> colorHexProperty().set(CColor.toHex(getValue())));
        setOnHidden(event -> colorHexProperty().set(CColor.toHex(getValue())));

        colorHexProperty().addListener((observable, oldValue, newValue) -> {
            String colorHex = newValue;
            int length = colorHex.length();

            switch (length) {
                case 1:
                    colorHex = CString.repeat(colorHex, 6);
                    break;
                case 3:
                    StringBuilder builder = new StringBuilder();
                    for (char c : colorHex.toCharArray()) {
                        builder.append(CString.repeat(c, 2));
                    }
                    colorHex = builder.toString();
                    break;
            }
            length = colorHex.length();

            if (length == 6 || length == 9) {
                setValue(Color.web(colorHex));
            }
        });
    }

    public StringProperty colorHexProperty() {
        return colorHexField.textProperty();
    }

    @Override
    protected Skin<?> createDefaultSkin() {
        return new CustomColorPickerSkin(this);
    }

    private class CustomColorPickerSkin extends ColorPickerSkin {

        public CustomColorPickerSkin(ColorPicker control) {
            super(control);
        }

        @Override
        protected Node getPopupContent() {
            // This is an instance of private API ColorPalette which extends Region
            Node colorPalette = super.getPopupContent();
            Region region = (Region) colorPalette;

            // This ColorPalette contains a VBox which contains the Hyperlink we want to remove.
            List<Node> nVBoxes = region.getChildrenUnmodifiable().stream().filter(e -> e instanceof VBox).collect(Collectors.toList());
            for (Node n : nVBoxes) {
                VBox vbox = (VBox) n;
                List<Node> hyperlinks = vbox.getChildren().stream().filter(e -> e instanceof Hyperlink).collect(Collectors.toList());

                vbox.getChildren().removeAll(hyperlinks); // Remove the hyperlink
                if (hyperlinks.size() > 0) {
                    vbox.setAlignment(Pos.CENTER);

                    vbox.getChildren().add(colorHexField);
                }
            }
            return colorPalette;
        }

    }

}

File: src/main/kotlin/info/meodinger/lpfx/component/CColorPicker.kt
package info.meodinger.lpfx.component

import info.meodinger.lpfx.util.color.toHex
import info.meodinger.lpfx.util.char.repeat

import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.geometry.Pos
import javafx.scene.Node
import javafx.scene.control.*
import javafx.scene.control.skin.ColorPickerSkin
import javafx.scene.input.KeyCode
import javafx.scene.layout.Region
import javafx.scene.layout.VBox
import javafx.scene.paint.Color
import java.util.stream.Collectors

/**
 * Author: Meodinger
 * Date: 2021/7/29
 * Location: info.meodinger.lpfx.component
 */
class CColorPicker : ColorPicker() {

    private val colorHexField = TextField()
    private val colorHexProperty = colorHexField.textProperty()

    init {
        colorHexField.textFormatter = TextFormatter<String> { change ->
            change.text = change.text.uppercase()
            if (change.isAdded) {
                val builder = StringBuilder()
                for (c in change.text.toCharArray()) {
                    if ((c in '0'..'9') || (c in 'A'..'F')) {
                        builder.append(c)
                    }
                }
                change.text = builder.toString()
            }
            change
        }
        colorHexField.onKeyReleased = EventHandler { event ->
            if (event.code == KeyCode.ENTER) {
                fireEvent(ActionEvent())
                hide()
            }
        }

        colorHexProperty.addListener { _, _, newValue ->
            when (newValue.length) {
                1 -> value = Color.web(newValue.repeat(6))
                2 -> value = Color.web(newValue.repeat(3))
                3 -> {
                    val builder = StringBuilder()
                    for (c in newValue.toCharArray()) {
                        builder.append(c.repeat(2))
                    }
                    value = Color.web(builder.toString())
                }
                6 -> value = Color.web(newValue)
            }
        }

        addEventHandler(ON_SHOWN) { colorHexProperty.set(value.toHex()) }
        addEventHandler(ON_HIDDEN) { colorHexProperty.set(value.toHex()) }
    }

    override fun createDefaultSkin(): Skin<*> {
        return object : ColorPickerSkin(this) {
            override fun getPopupContent(): Node {
                // This is an instance of private API ColorPalette which extends Region
                val colorPalette: Region = super.getPopupContent() as Region

                // This ColorPalette contains a VBox which contains the Hyperlink we want to remove.
                val nVBoxes: List<Node> = colorPalette.childrenUnmodifiable
                    .stream()
                    .filter { e -> e is VBox }
                    .collect(Collectors.toList())
                for (node in nVBoxes) {
                    val vbox: VBox = node as VBox
                    val hyperlinks: List<Node> = vbox.children
                        .stream()
                        .filter { e -> e is Hyperlink }
                        .collect(Collectors.toList())
                    if (hyperlinks.isNotEmpty()) {
                        vbox.children.removeAll(hyperlinks) // Remove the hyperlink
                        vbox.alignment = Pos.CENTER
                        vbox.children.add(colorHexField)
                    }
                }
                return colorPalette
            }
        }
    }
}